File: src/main/java/de/aquafun3d/bingo/utils/scoreboards/Scoreboards.java
package de.aquafun3d.bingo.utils.scoreboards;

import net.kyori.adventure.text.Component;
import org.bukkit.Bukkit;
import org.bukkit.ChatColor;
import org.bukkit.entity.Player;
import org.bukkit.scoreboard.Criteria;
import org.bukkit.scoreboard.Objective;
import org.bukkit.scoreboard.Scoreboard;
import org.bukkit.scoreboard.Team;

import java.util.HashMap;
import java.util.UUID;

public class Scoreboards implements IScoreboards{

    private final HashMap<UUID, Scoreboard> _playerScoreboards = new HashMap<>();
    private final HashMap<UUID, Objective> _playerObjectives = new HashMap<>();
    private final Scoreboard _defaultBoard = Bukkit.getScoreboardManager().getNewScoreboard();

    public Scoreboards(){
        initDefaultBoard();
    }

    public void initPlayerScorebaord(Player player){
        UUID uuid = player.getUniqueId();
        if(!_playerScoreboards.containsKey(uuid)){
            _playerScoreboards.put(uuid, _defaultBoard);
        }
        player.setScoreboard(_playerScoreboards.get(uuid));
        _playerObjectives.put(uuid,player.getScoreboard().getObjective("bingo"));
    }

    private void addTeam(Scoreboard board, String teamname, String prefix){
        board.registerNewTeam(teamname);
        board.getTeam(teamname).prefix(Component.text(prefix));
    }

    private void initScore(String teamname, String value, int score){
            Objective obj = _defaultBoard.getObjective("bingo");
            Team team = _defaultBoard.getTeam(teamname);
            team.prefix(Component.text(value));
            obj.getScore(value).setScore(score);
    }

    public void updateScore(Player player, String teamname, String value, int score, boolean remove){
        UUID uuid = player.getUniqueId();
        Scoreboard board = _playerScoreboards.get(uuid);
        Objective obj = _playerObjectives.get(uuid);
        board.resetScores(board.getTeam(teamname).prefix().toString());
        if(!remove) {
            board.getTeam(teamname).prefix(Component.text(value));
            obj.getScore(value).setScore(score);
        }
    }

    public HashMap<UUID, Scoreboard> getPlayerBoards() {
        return this._playerScoreboards;
    }

    private void initDefaultBoard() {
        addTeam(_defaultBoard, "spec", ChatColor.GRAY + "[Spec] ");
        addTeam(_defaultBoard, "white", ChatColor.WHITE + "[#1] ");
        addTeam(_defaultBoard, "orange", ChatColor.GOLD + "[#2] ");
        addTeam(_defaultBoard, "magenta", ChatColor.LIGHT_PURPLE + "[#3] ");
        addTeam(_defaultBoard, "lblue", ChatColor.AQUA + "[#4] ");
        addTeam(_defaultBoard, "yellow", ChatColor.YELLOW + "[#5] ");
        addTeam(_defaultBoard, "lime", ChatColor.GREEN + "[#6] ");
        addTeam(_defaultBoard, "pink", ChatColor.RED + "[#7] ");
        addTeam(_defaultBoard, "gray", ChatColor.DARK_GRAY + "[#8] ");
        addTeam(_defaultBoard, "lgray", ChatColor.GRAY + "[#9] ");
        addTeam(_defaultBoard, "cyan", ChatColor.DARK_AQUA + "[#10] ");
        addTeam(_defaultBoard, "purple", ChatColor.DARK_PURPLE + "[#11] ");
        addTeam(_defaultBoard, "blue", ChatColor.BLUE + "[#12] ");
        addTeam(_defaultBoard, "brown", ChatColor.GOLD + "[#13] ");
        addTeam(_defaultBoard, "green", ChatColor.DARK_GREEN + "[#14] ");
        addTeam(_defaultBoard, "red", ChatColor.DARK_RED + "[#15] ");
        addTeam(_defaultBoard, "black", ChatColor.DARK_GRAY + "[#16] ");
        addTeam(_defaultBoard, "firstplace", ChatColor.GOLD + "1. Place: ");
        addTeam(_defaultBoard, "place", ChatColor.GOLD + "Your Team: ");
        addTeam(_defaultBoard, "blank", " ");
        addTeam(_defaultBoard, "itemcount", "X items left");

        _defaultBoard.registerNewObjective("bingo", Criteria.DUMMY, Component.text("bingo"));

        initScore("firstplace", ChatColor.GOLD + "1. Place: ", 13);
        initScore("blank", " ", 11);
        initScore("blank", " ", 14);
        initScore("place", ChatColor.GOLD + "Your Team: ", 12);
        initScore("itemcount", "X items left", 10);
    }
}

File: src/main/java/de/aquafun3d/bingo/utils/scoreboards/Scoreboards.kt
package de.aquafun3d.bingo.utils.scoreboards

import net.kyori.adventure.text.Component
import net.kyori.adventure.text.format.NamedTextColor
import org.bukkit.Bukkit
import org.bukkit.entity.Player
import org.bukkit.scoreboard.Scoreboard
import java.util.*

class Scoreboards : IScoreboards {
    private val _playerBoards = mutableMapOf<UUID, Scoreboard>()
    private val _defaultBoard = Bukkit.getScoreboardManager().newScoreboard

    init {
        initDefaultBoard()
    }

    override fun initPlayerScorebaord(player: Player) {
        val uuid = player.uniqueId
        if (!_playerBoards.containsKey(uuid)) {
            _playerBoards[uuid] = _defaultBoard
        }
        player.scoreboard = _playerBoards[uuid]!!
    }

    private fun addTeam(board: Scoreboard, teamname: String, prefix: Component) {
        board.registerNewTeam(teamname)
        board.getTeam(teamname)!!.prefix(prefix)
    }

    private fun initDefaultBoard() {
        addTeam(_defaultBoard, "spec", Component.text("[Spec] ", NamedTextColor.GRAY))
        addTeam(_defaultBoard, "white", Component.text("[#1] ",NamedTextColor.WHITE))
        addTeam(_defaultBoard, "orange", Component.text("[#2] ", NamedTextColor.GOLD))
        addTeam(_defaultBoard, "magenta", Component.text("[#3] ", NamedTextColor.LIGHT_PURPLE))
        addTeam(_defaultBoard, "lightblue", Component.text("[#4] ", NamedTextColor.AQUA))
        addTeam(_defaultBoard, "yellow", Component.text("[#5] ", NamedTextColor.YELLOW))
        addTeam(_defaultBoard, "lime", Component.text("[#6] ",NamedTextColor.GREEN))
        addTeam(_defaultBoard, "pink", Component.text("[#7] ", NamedTextColor.RED))
        addTeam(_defaultBoard, "gray", Component.text("[#8] ", NamedTextColor.DARK_GRAY))
        addTeam(_defaultBoard, "lightgray", Component.text("[#9] ", NamedTextColor.GRAY))
        addTeam(_defaultBoard, "cyan", Component.text("[#10] ", NamedTextColor.DARK_AQUA))
        addTeam(_defaultBoard, "purple", Component.text("[#11] ", NamedTextColor.DARK_PURPLE))
        addTeam(_defaultBoard, "blue", Component.text("[#12] ", NamedTextColor.BLUE))
        addTeam(_defaultBoard, "brown", Component.text("[#13] ", NamedTextColor.GOLD))
        addTeam(_defaultBoard, "green", Component.text("[#14] ", NamedTextColor.DARK_GREEN))
        addTeam(_defaultBoard, "red", Component.text("[#15] ", NamedTextColor.DARK_RED))
        addTeam(_defaultBoard, "black", Component.text("[#16] ", NamedTextColor.DARK_GRAY))
        addTeam(_defaultBoard, "blank", Component.text(" "))
    }

    override fun getPlayerboards(): Map<UUID, Scoreboard>{
        return _playerBoards
    }
}
